### PR TITLE
[NR-343531] Workflows for lambda code and templates deployment

### PR DESCRIPTION
**Pull request by ahegdeNR**
In github release workflow add step to build and deploy lambda code to all regions S3 bucket and push it to logint's AWS account along with templates. 

Note -> The S3 buckets created and added in the workflow are test buckets. Once the IAM role for Jim Ruppert account is created the workflow will have production buckets added. 

Tests run:
1. deployed firehose-lambda-template in us-east-2. Stack name -> anusha-ci-cd-test-lambda-firehose-template. Logs populated in nr. 
2. deployed lambda-template in eu-west-1. Stack name -> anusha-ci-cd-test-lambda-template. Logs populated in nr 

Screenshot 2024-11-28 at 7 15 54 PM

Check the workflow run here successfully -> https://github.com/newrelic/aws-unified-lambda/pull/1/checks?check_run_id=33904514146

### CODE REVIEW

File: .github/workflows/release.yaml [hunk 1]
```diff
@@ -26,15 +26,49 @@ jobs:
           role-to-assume: ${{ secrets.AWS_SAR_PUBLISHER_ROLE }}
           aws-region: us-east-2
 
+      - name: Build and Package Go executable
+        run: |
+          cd src
```
Comment (voorepreethi) [inline .github/workflows/release.yaml:31]: Include 'go mod tidy' for downloading all the dependencies if any

File: .github/workflows/release.yaml [hunk 2]
```diff
@@ -26,15 +26,49 @@ jobs:
...
+            "logging-lambda-firehose-metric-stream.yaml"
+            "lambda-template.yaml"
+            "logging-lambda-firehose-template.yaml"
+          )
+          
+          for TEMPLATE_FILE in "${TEMPLATES[@]}"; do
+            BASE_NAME=$(basename "$TEMPLATE_FILE" .yaml)
+            BUILD_DIR="$BUILD_DIR_BASE/$BASE_NAME"
+          
+            sam build --template-file "$TEMPLATE_FILE" --build-dir "$BUILD_DIR"
```
Comment (voorepreethi) [inline .github/workflows/release.yaml:70]: add error handling incase of failure of uplaoding

File: .github/workflows/release.yaml [hunk 3]
```diff
@@ -26,15 +26,49 @@ jobs:
...
+          cd src
+          GOOS=linux GOARCH=amd64 go build -o bootstrap main.go
+          zip -r ../new-relic-log-forwarder.zip .
+
+      - name: Upload lambda code to S3 buckets in all AWS region
+        env:
+          file_to_upload: new-relic-log-forwarder.zip
+        run: |
+          regions=("us-east-1" "us-east-2" "eu-west-1" "eu-west-2")
+          #"us-west-1" "us-west-2" "af-south-1" "ap-east-1" "ap-south-2" "ap-southeast-3" "ap-southeast-5" "ap-southeast-4" "ap-south-1" "ap-northeast-3" "ap-northeast-2" "ap-southeast-1" "ap-southeast-2" "ap-northeast-1" "ca-central-1" "ca-west-1" "eu-central-1" "eu-south-1" "eu-west-3" "eu-south-2" "eu-north-1" "eu-central-2" "me-south-1" "me-central-1" "il-central-1" "sa-east-1")
```
[the remark this answers is not in the document]
Comment (maya-jha) [reply .github/workflows/release.yaml:40]: you can parameterize this prefix also: new-relic-log-forwarder-folder

File: .github/workflows/release.yaml [hunk 4]
```diff
@@ -26,15 +26,49 @@ jobs:
...
+        run: |
+          cd src
+          GOOS=linux GOARCH=amd64 go build -o bootstrap main.go
+          zip -r ../new-relic-log-forwarder.zip .
+
+      - name: Upload lambda code to S3 buckets in all AWS region
+        env:
+          file_to_upload: new-relic-log-forwarder.zip
+        run: |
+          regions=("us-east-1" "us-east-2" "eu-west-1" "eu-west-2")
```
[the remark this answers is not in the document]
Comment (maya-jha) [reply .github/workflows/release.yaml:39]: if you are adding other regions later, add a TODO comment here.

File: .github/workflows/release.yaml [hunk 5]
```diff
@@ -26,15 +26,49 @@ jobs:
...
+          #"us-west-1" "us-west-2" "af-south-1" "ap-east-1" "ap-south-2" "ap-southeast-3" "ap-southeast-5" "ap-southeast-4" "ap-south-1" "ap-northeast-3" "ap-northeast-2" "ap-southeast-1" "ap-southeast-2" "ap-northeast-1" "ca-central-1" "ca-west-1" "eu-central-1" "eu-south-1" "eu-west-3" "eu-south-2" "eu-north-1" "eu-central-2" "me-south-1" "me-central-1" "il-central-1" "sa-east-1")
+          for region in "${regions[@]}"; do
+            bucket_name="unified-lambda-logging-${region}"
```
Comment (maya-jha) [inline .github/workflows/release.yaml:42]: nit: Does this bucket exist in logint account? We could have added test also in name, but as this is a temporary thing we can keep it as it is.

File: .github/workflows/release.yaml [hunk 6]
```diff
@@ -26,15 +26,49 @@ jobs:
...
-        run: sam build --template-file template.yaml --region us-east-2
-
-      - name: Package SAM Application
-        run: sam package --s3-bucket unified-lambda-cft-1 --output-template-file packaged.yaml --region us-east-2
-# should we include retries as part of uploading?
-      - name: Upload CloudFormation Template to S3
-        run: aws s3 cp packaged.yaml s3://unified-lambda-serverless-1/packaged.yaml
+      - name: Build and Package SAM Application and Upload to S3
+        env:
+          bucket_name: anusha-test-ansuha
```
Comment (maya-jha) [inline .github/workflows/release.yaml:52]: parametrize this also, and add a TODO comment to change it when jim rupert account access is working.

File: .github/workflows/release.yaml [hunk 7]
```diff
@@ -26,15 +26,49 @@ jobs:
...
+            "logging-firehose-metric-polling.yaml"
+            "logging-firehose-metric-stream.yaml"
+            "logging-lambda-firehose-metric-polling.yaml"
+            "logging-lambda-firehose-metric-stream.yaml"
+            "lambda-template.yaml"
+            "logging-lambda-firehose-template.yaml"
+          )
+          
+          for TEMPLATE_FILE in "${TEMPLATES[@]}"; do
+            BASE_NAME=$(basename "$TEMPLATE_FILE" .yaml)
```
Comment (maya-jha) [inline .github/workflows/release.yaml:67]: why are we adding `basename` here?

File: .github/workflows/release.yaml [hunk 8]
```diff
@@ -26,15 +26,49 @@ jobs:
...
+            BUILD_DIR="$BUILD_DIR_BASE/$BASE_NAME"
+          
+            sam build --template-file "$TEMPLATE_FILE" --build-dir "$BUILD_DIR"
+            sam package --s3-bucket "$bucket_name" --template-file "$BUILD_DIR/template.yaml" --output-template-file "$BUILD_DIR/$TEMPLATE_FILE" --region us-east-1
```
Comment (maya-jha) [inline .github/workflows/release.yaml:71]: parametrize `us-east-1` also here and we should use us-east-2 to keep it same as our prod bucket to minimize changes later.

**Review by maya-jha — CHANGES_REQUESTED** [review]
1. How are you going to test it? Do you have to create a manual release in github after PR is merged?
2. Does this contain latest code from our internal repo, maybe confirm with Preeti?

File: .github/workflows/release.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (voorepreethi) [reply .github/workflows/release.yaml:39]: It contains latest code in master as of Friday

File: .github/workflows/release.yaml (hunk 6 above)
Comment (ahegdeNR) [reply .github/workflows/release.yaml:52]: @maya-jha it's already parameterized as it's under env_variables. Let me know if you're looking for anything else.

File: .github/workflows/release.yaml [hunk 9]
```diff
@@ -26,15 +26,49 @@ jobs:
...
+
+      - name: Upload lambda code to S3 buckets in all AWS region
+        env:
+          file_to_upload: new-relic-log-forwarder.zip
+        run: |
+          regions=("us-east-1" "us-east-2" "eu-west-1" "eu-west-2")
+          #"us-west-1" "us-west-2" "af-south-1" "ap-east-1" "ap-south-2" "ap-southeast-3" "ap-southeast-5" "ap-southeast-4" "ap-south-1" "ap-northeast-3" "ap-northeast-2" "ap-southeast-1" "ap-southeast-2" "ap-northeast-1" "ca-central-1" "ca-west-1" "eu-central-1" "eu-south-1" "eu-west-3" "eu-south-2" "eu-north-1" "eu-central-2" "me-south-1" "me-central-1" "il-central-1" "sa-east-1")
+          for region in "${regions[@]}"; do
+            bucket_name="unified-lambda-logging-${region}"
+            aws s3 cp "$file_to_upload" "s3://$bucket_name/new-relic-log-forwarder-folder/" --region "$region"
```
[the remark this answers is not in the document]
Comment (ahegdeNR) [reply .github/workflows/release.yaml:43]: added matrix to the step which deploys code and runs in parallel in different VMs

File: .github/workflows/release.yaml (hunk 1 above)
Comment (ahegdeNR) [reply .github/workflows/release.yaml:28]: updated

File: .github/workflows/release.yaml (hunk 6 above)
Comment (ahegdeNR) [reply .github/workflows/release.yaml:102]: updated

File: .github/workflows/release.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (ahegdeNR) [reply .github/workflows/release.yaml]: updated

File: .github/workflows/release.yaml (hunk 9 above)
[the remark this answers is not in the document]
Comment (ahegdeNR) [reply .github/workflows/release.yaml]: updated

File: .github/workflows/release.yaml (hunk 8 above)
Comment (ahegdeNR) [reply .github/workflows/release.yaml]: updated

File: .github/workflows/release.yaml (hunk 7 above)
Comment (ahegdeNR) [reply .github/workflows/release.yaml:69]: This is just to separate all the template files and adding them to specific directory and then uploading to bucket

**Pull request closed without merging**